### DISCUSSION TITLE
Testing AsyncNotifierProvider: unexpected loading

### DISCUSSION THREAD

**Discussion by YSDC** [rrousselGit/riverpod#2295] [Q&A]
Hello everyone,
I was implementing some unit tests and the time came to test the Async provider. I followed the 'famous' tutorial from Andrea [here](https://codewithandrea.com/articles/unit-test-async-notifier-riverpod/). The thing is, I have an unexpected state _AsyncLoading_ that I never pass. Let me show you some details.

Here is my provider:

```
@riverpod
class UserController extends _$UserController {
  @override
  FutureOr<User> build() async {
    var secureStorage = ref.watch(secureStorageProvider);
    User user = await secureStorage.getUserDetails();
    state = AsyncData(user);
    return user;
  }

  void updateUser(User user) {
    state = AsyncData(user);
  }
}
```

And here is part of my test class:
```
     ProviderContainer container = makeContainer();
      final UserRepository userRepository = container.read(userRepositoryProvider);
      final controller = container.read(userControllerProvider.notifier);

      final listener = Listener<AsyncValue<User>>();
      container.listen(userControllerProvider, listener, fireImmediately: true);

      verify(() => listener(null, AsyncData<User>(user)));
```

I was expecting to get a null => AsyncData but in reality, I have a null => AsyncLoading
` No matching calls. All calls: Listener<AsyncValue<User>>.call(null, AsyncLoading<User>())`

If you look, I only called the build method, implicitly, by initializing my provider.

Now look at the stack when I call the _UpdateUser_ method of my provider:

```
 Matching call #0 not found. All calls: 
 Listener<AsyncValue<User>>.call(null, AsyncLoading<User>()), 
 Listener<AsyncValue<User>>.call(AsyncLoading<User>(), AsyncData<User>(value: Instance of 'User')), 
 Listener<AsyncValue<User>>.call(AsyncData<User>(value: Instance of 'User'), AsyncData<User>(value: Instance of 'User')), 
 Listener<AsyncValue<User>>.call(AsyncData<User>(value: Instance of 'User'), AsyncData<User>(value: Instance of 'User'))
```

**Comment by rrousselGit** [comment]
Your code is async. The loading is there naturally
